cli/azd/internal/cmd/publish.go

```diff
@@ -0,0 +1,373 @@
+				func(packageProgress project.ServiceProgress) {
+					progressMessage := fmt.Sprintf("Packaging service %s (%s)", svc.Name, packageProgress.Message)
+					pa.console.ShowSpinner(ctx, progressMessage, input.Step)
+				},
+				func(progress *async.Progress[project.ServiceProgress]) (*project.ServicePackageResult, error) {
+					return pa.serviceManager.Package(ctx, svc, nil, progress, nil)
+				},
+			)
+
+			if err != nil {
+				pa.console.StopSpinner(ctx, stepMessage, input.StepFailed)
+				return nil, err
+			}
+		}
+
+		publishResult, err := async.RunWithProgress(
+			func(publishProgress project.ServiceProgress) {
+				progressMessage := fmt.Sprintf("Publishing service %s (%s)", svc.Name, publishProgress.Message)
+				pa.console.ShowSpinner(ctx, progressMessage, input.Step)
+			},
+			func(progress *async.Progress[project.ServiceProgress]) (*project.ServicePublishResult, error) {
+				return pa.serviceManager.Publish(ctx, svc, packageResult, progress, publishOptions)
+			},
+		)
+
+		// clean up for packages automatically created in temp dir
+		if pa.flags.FromPackage == "" && strings.HasPrefix(packageResult.PackagePath, os.TempDir()) {
+			if err := os.RemoveAll(packageResult.PackagePath); err != nil {
+				log.Printf("failed to remove temporary package: %s : %s", packageResult.PackagePath, err)
+			}
+		}
+
+		pa.console.StopSpinner(ctx, stepMessage, input.GetStepResultFormat(err))
+		if err != nil {
+			return nil, err
+		}
+
+		publishResults[svc.Name] = publishResult
+		pa.console.MessageUxItem(ctx, publishResult)
+	}
+
+	if pa.formatter.Kind() == output.JsonFormat {
+		publishResult := PublishResult{
+			Timestamp: time.Now(),
+			Services:  publishResults,
+		}
+
+		if fmtErr := pa.formatter.Format(publishResult, pa.writer, nil); fmtErr != nil {
+			return nil, fmt.Errorf("publish result could not be displayed: %w", fmtErr)
+		}
+	}
+
+	return &actions.ActionResult{
+		Message: &actions.ResultMessage{
+			Header: fmt.Sprintf("Your application was published in %s.",
+				ux.DurationAsText(since(startTime))),
+		},
+	}, nil
+}
+
+func GetCmdPublishHelpDescription(*cobra.Command) string {
+	return generateCmdHelpDescription(
+		"Publish a service to a container registry.",
+		[]string{
+			formatHelpNote("Supports Container App services only."),
+			formatHelpNote(
+				//nolint:lll
+				"Target registry set by AZURE_CONTAINER_REGISTRY_ENDPOINT environment variable or docker.registry in azure.yaml."),
+			formatHelpNote(
+				//nolint:lll
+				"Use '--from-package' to publish an existing container image, otherwise azd automatically packages the container image before publishing."),
+		})
+}
+
+func GetCmdPublishHelpFooter(*cobra.Command) string {
+	return generateCmdHelpSamplesBlock(map[string]string{
+		"Publish all services in the current project.": output.WithHighLightFormat(
+			"azd publish --all",
+		),
+		"Publish the service named 'api'.": output.WithHighLightFormat(
+			"azd publish api",
+		),
```

Comment:
I am thinking out loud and sitting on this, wondering how natural this feels. I feel like it would be nice to have separate atomic commands, that `azd publish api` wouldn't "over-emcompass" the idea of packaging... However, what you have is indeed consistent with our current model.